stl/inc/format

```diff
@@ -0,0 +1,2734 @@
+                _Handler._On_text(_Begin, _ClosingCurl);
+
+                // skip over the second closing curl
+                _Begin = _ClosingCurl + 1;
+            }
+
+            // We are done, there were no replacement fields.
+            if (_OpeningCurl == _End) {
+                return;
+            }
+        }
+        // Parse the replacement field starting at _OpeningCurl and ending sometime before _End.
+        _Begin = _Parse_replacement_field(_OpeningCurl, _End, _Handler);
+    }
+}
+
+template <class _CharT>
+struct _Basic_format_specs {
+    int _Width           = 0;
+    int _Precision       = -1;
+    char _Type           = '\0';
+    _Align _Alignment    = _Align::_None;
+    _Sign _Sgn           = _Sign::_None;
+    bool _Alt            = false;
+    bool _Localized      = false;
+    bool _Leading_zero   = false;
+    uint8_t _Fill_length = 1;
+    // At most one codepoint (so one char32_t or four utf-8 char8_t).
+    _CharT _Fill[4 / sizeof(_CharT)] = {_CharT{' '}};
+};
+
+// Adds width and precision references to _Basic_format_specs.
+// this is required for std::formatter implementations because we must
+// parse the format specs without having access to the format args (via a format context)
+template <class _CharT>
+struct _Dynamic_format_specs : _Basic_format_specs<_CharT> {
+    int _Dynamic_width_index     = -1;
+    int _Dynamic_precision_index = -1;
+};
+
+// Model of _Parse_spec_callbacks that fills a _Basic_format_specs with the parsed data.
+template <class _CharT>
+class _Specs_setter {
+public:
+    explicit constexpr _Specs_setter(_Basic_format_specs<_CharT>& _Specs_) : _Specs(_Specs_) {}
+
+    constexpr void _On_align(_Align _Aln) {
+        _Specs._Alignment = _Aln;
+    }
+
+    constexpr void _On_fill(basic_string_view<_CharT> _Sv) {
+        if (_Sv.size() > _STD size(_Specs._Fill)) {
+            throw format_error("Invalid fill (too long).");
+        }
+
+        const auto _Pos = _STD _Copy_unchecked(_Sv._Unchecked_begin(), _Sv._Unchecked_end(), _Specs._Fill);
+        _STD fill(_Pos, _STD end(_Specs._Fill), _CharT{});
+        _Specs._Fill_length = static_cast<uint8_t>(_Sv.size());
+    }
+
+    constexpr void _On_sign(_Sign _Sgn) {
+        _Specs._Sgn = _Sgn;
+    }
+
+    constexpr void _On_hash() {
+        _Specs._Alt = true;
+    }
+
+    constexpr void _On_zero() {
+        _Specs._Leading_zero = true;
+    }
+
+    constexpr void _On_width(int _Width) {
+        _Specs._Width = _Width;
+    }
+
+    constexpr void _On_precision(int _Precision) {
+        _Specs._Precision = _Precision;
+    }
+
+    constexpr void _On_localized() {
+        _Specs._Localized = true;
+    }
+
+    constexpr void _On_type(_CharT _Type) {
+        // performance note: this could be optimized to one comparison by
+        // first casting to unsigned int (the negative values will be 128-255)
+        if (_Type < 0 || _Type > (numeric_limits<signed char>::max)()) {

```

Comment:
Should we apply this optimization?
```suggestion
        if (static_cast<unsigned int>(_Type) >= 128) { // negative values start at 128
```

Reply:
I recall asking this in another CR, and someone showed me that compilers already perform this optimization. Here, even if the character type is signed with a range that could be too negative or too positive, we still get a single comparison: https://godbolt.org/z/PPj16eeqT

Accordingly, I think we should simply remove this performance note and keep the clear code.